### PR TITLE
[ISSUE-792] Fix volume creation with corrupted FS

### LINKED ISSUE #792: CSI pv have segments from previous deployments when used hdd storage class
**Describe the bug**
On PrepareVolume CSI return success after partition creation and existed FS:
```
Jan 10 12:27:45.350 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa] Create partition {/dev/sdl 1 gpt CSI 603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa false} on device /dev/sdl and set UUID
Jan 10 12:27:45.350 [DEBU] [PartitionOperations] [PreparePartition] [603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa] Processing for partition utilwrappers.Partition{Device:"/dev/sdl", Name:"", Num:"1", TableType:"gpt", Label:"CSI", PartUUID:"603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa", Ephemeral:false}
Jan 10 12:27:45.405 [DEBU] [Executor] [partprobe -d -s /dev/sdl] [55.019465ms] [55019465] stdout: /dev/sdl: gpt partitions 1
Jan 10 12:27:45.407 [DEBU] [Executor] [sgdisk /dev/sdl --info=1] [2.178954ms] [2178954] stdout: Partition GUID code: 0FC63DAF-8483-4772-8E79-3D69D8477DE4 (Linux filesystem)
Jan 10 12:27:45.407 [DEBU] [PartitionOperations] [SearchPartName] [603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa] Search partition number for device /dev/sdl and uuid 603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa
Jan 10 12:27:45.419 [DEBU] [Executor] [blockdev --rereadpt -v /dev/sdl] [11.895385ms] [11895385] stdout: 
Jan 10 12:27:45.550 [INFO] [Controller] [drive/Reconcile] [d1a7e839-9263-4b2c-8bfb-e1c82e5e5a46] Drive changed: &{{ } {d1a7e839-9263-4b2c-8bfb-e1c82e5e5a46 /apis/csi-baremetal.dell.com/v1/drives/d1a7e839-9263-4b2c-8bfb-e1c82e5e5a46 31fad433-8192-4572-97f9-939ee294996c 7303780 4 2021-12-27 10:52:57 +0000 UTC <nil> <nil> map[app:csi-baremetal app.kubernetes.io/name:csi-baremetal] map[] [] [] [{node Update csi-baremetal.dell.com/v1 2022-01-10 12:18:11 +0000 UTC FieldsV1 &FieldsV1{Raw:*[123 34 102 58 109 101 116 97 100 97 116 97 34 58 123 34 102 58 108 97 98 101 108 115 34 58 123 34 46 34 58 123 125 44 34 102 58 97 112 112 34 58 123 125 44 34 102 58 97 112 112 46 107 117 98 101 114 110 101 116 101 115 46 105 111 47 110 97 109 101 34 58 123 125 125 125 44 34 102 58 115 112 101 99 34 58 123 34 46 34 58 123 125 44 34 102 58 70 105 114 109 119 97 114 101 34 58 123 125 44 34 102 58 72 101 97 108 116 104 34 58 123 125 44 34 102 58 78 111 100 101 73 100 34 58 123 125 44 34 102 58 80 73 68 34 58 123 125 44 34 102 58 80 97 116 104 34 58 123 125 44 34 102 58 83 101 114 105 97 108 78 117 109 98 101 114 34 58 123 125 44 34 102 58 83 105 122 101 34 58 123 125 44 34 102 58 83 108 111 116 34 58 123 125 44 34 102 58 83 116 97 116 117 115 34 58 123 125 44 34 102 58 84 121 112 101 34 58 123 125 44 34 102 58 85 85 73 68 34 58 123 125 44 34 102 58 85 115 97 103 101 34 58 123 125 44 34 102 58 86 73 68 34 58 123 125 125 125],}}]} {d1a7e839-9263-4b2c-8bfb-e1c82e5e5a46 ATA TOSHIBA MG06ACA8 90J0A1ZEF1QF GOOD HDD 8001000000000 ONLINE IN_USE f3395e78-6364-4122-8eee-554308b64e19 /dev/sdl 22 GA08 0 false false {} [] 0}}
Jan 10 12:27:48.424 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa] Partition was created successfully &{Device:/dev/sdl Name:1 Num:1 TableType:gpt Label:CSI PartUUID:603aa6c7-1d88-4518-b7e0-c2fb5d2e84fa Ephemeral:false}
Jan 10 12:27:48.426 [DEBU] [Executor] [lsblk /dev/sdl1 --output FSTYPE --noheadings] [1.461078ms] [1461078] stdout: xfs
Jan 10 12:27:48.426 [WARN] [FSOperations] [CreateFSIfNotExist] FS on /dev/sdl1 with type xfs is already exist, skip creating
```

**Environment (please complete the following information):**
 - OS: RHEL
 - Platform: Openshift

**Expected behavior**
RCA - CSI uses previously created FS on the disk. Need to set FS UUID and check it to avoid such issue. Need to review disk cleanup procedure as well.

### PR DESCRIPTION
**Pull request by safronovD**
### Resolves #792 

- Create FS with volume UUID
- Fail FS creation if UUID is wrong

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

### Success scenario
Create volume
```
user@user-vm ~/g/s/g/dell> ka csi-baremetal/test/app/myn.yaml
statefulset.apps/web created

user@user-vm ~/g/s/g/dell> kgo volume
NAME                                       SIZE        STORAGE CLASS   HEALTH   CSI_STATUS   OP_STATUS   USAGE    TYPE   LOCATION                               NODE
pvc-97062541-885c-49fa-846f-1e9a43816939   105906176   HDD             GOOD     PUBLISHED    OPERATIVE   IN_USE   xfs    73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d   0e7f8009-62bd-45f4-a66e-38b09566ef66
```
Check FS UUID
```
user@user-vm ~/g/s/g/dell> kgo drive 73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d
NAME                                   SIZE        TYPE   HEALTH   STATUS   USAGE    SYSTEM   PATH          SERIAL NUMBER        NODE                                   SLOT
73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d   105906176   HDD    GOOD     ONLINE   IN_USE            /dev/loop17   LOOPBACK2138684047   0e7f8009-62bd-45f4-a66e-38b09566ef66   

user@user-vm ~/g/s/g/dell> lsblk /dev/loop17 -o name,mountpoint,size,uuid
NAME       MOUNTPOINT  SIZE UUID
loop17                 101M 
└─loop17p1             100M 97062541-885c-49fa-846f-1e9a43816939
```

### Corrupted FS
Create FS
```
user@user-vm ~/g/s/g/dell [4]> sudo sgdisk  /dev/loop17 -o
Creating new GPT entries in memory.
The operation has completed successfully.
user@user-vm ~/g/s/g/dell> sudo sgdisk -n 1:0:0 -c 1:CSI /dev/loop17
Setting name!
partNum is 0
The operation has completed successfully.
user@user-vm ~/g/s/g/dell> sudo mkfs.ext3 /dev/loop17p1 -U cb789d13-5e18-4e1e-866d-ead0d957d2b3
mke2fs 1.45.5 (07-Jan-2020)
Creating filesystem with 25595 4k blocks and 25600 inodes

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (1024 blocks): done
Writing superblocks and filesystem accounting information: done

user@user-vm ~/g/s/g/dell> lsblk /dev/loop17 -o name,mountpoint,size,uuid
NAME       MOUNTPOINT  SIZE UUID
loop17                 101M 
└─loop17p1             100M cb789d13-5e18-4e1e-866d-ead0d957d2b3
```

Delete partition **without wiping FS**
```
user@user-vm ~/g/s/g/dell> sudo sgdisk -d 1 /dev/loop17
The operation has completed successfully.
user@user-vm ~/g/s/g/dell> sudo wipefs -af /dev/loop17
/dev/loop17: 8 bytes were erased at offset 0x00000200 (gpt): 45 46 49 20 50 41 52 54
/dev/loop17: 8 bytes were erased at offset 0x064ffe00 (gpt): 45 46 49 20 50 41 52 54
/dev/loop17: 2 bytes were erased at offset 0x000001fe (PMBR): 55 aa
user@user-vm ~/g/s/g/dell> lsblk /dev/loop17 -o name,mountpoint,size,uuid
NAME   MOUNTPOINT  SIZE UUID
loop17             101M 
user@user-vm ~/g/s/g/dell [1]> sudo fdisk --list /dev/loop17
Disk /dev/loop17: 101 MiB, 105906176 bytes, 206848 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

Create sts and check
```
user@user-vm ~/g/s/g/dell> kg volume
NAME                                       SIZE        STORAGE CLASS   HEALTH   CSI_STATUS   LOCATION                               NODE
pvc-631b3a1e-9183-4e1f-969f-356688736996   105906176   HDD             GOOD     CREATED      6360144b-5dae-4358-86c1-98ee5f4b6488   0e7f8009-62bd-45f4-a66e-38b09566ef66
pvc-7ae3a05a-696b-44d7-b4ef-baf801394c7e   105906176   HDD             GOOD     CREATED      1ecf248e-757a-44d5-b007-3f718b5724ab   0e7f8009-62bd-45f4-a66e-38b09566ef66
pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978   105906176   HDD             GOOD     FAILED       73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d   0e7f8009-62bd-45f4-a66e-38b09566ef66

user@user-vm ~/g/s/g/dell> kgo drive 73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d
NAME                                   SIZE        TYPE   HEALTH   STATUS   USAGE    SYSTEM   PATH          SERIAL NUMBER        NODE                                   SLOT
73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d   105906176   HDD    GOOD     ONLINE   IN_USE            /dev/loop17   LOOPBACK2138684047   0e7f8009-62bd-45f4-a66e-38b09566ef66   

user@user-vm ~/g/s/g/dell> lsblk /dev/loop17 -o name,mountpoint,size,uuid
NAME       MOUNTPOINT  SIZE UUID
loop17                 101M 
└─loop17p1             100M cb789d13-5e18-4e1e-866d-ead0d957d2b3
```

Node logs
```
Mar 25 08:55:38.438 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] Processing for volume {Id:pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978 Location:73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d LocationType:DRIVE StorageClass:HDD NodeId:0e7f8009-62bd-45f4-a66e-38b09566ef66 Owners:[] Size:105906176 Mode:FS Type:xfs Health:GOOD OperationalStatus:OPERATIVE CSIStatus:CREATING Usage:IN_USE Ephemeral:false XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
Mar 25 08:55:38.513 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] Search device file for drive with S/N LOOPBACK2138684047
Mar 25 08:55:38.513 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] Create partition {/dev/loop17  1 gpt CSI e91c07cf-60ef-4c3d-bdff-58e30594b978 false} on device /dev/loop17 and set UUID
Mar 25 08:55:47.521 [INFO] [DriveProvisioner] [PrepareVolume] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] Partition was created successfully &{Device:/dev/loop17 Name:p1 Num:1 TableType:gpt Label:CSI PartUUID:e91c07cf-60ef-4c3d-bdff-58e30594b978 Ephemeral:false}
Mar 25 08:55:47.546 [ERRO] [VolumeManager] [prepareVolume] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] [/home/user/go_projects/src/github.com/dell/csi-baremetal/pkg/node/volumemgr.go:468] [prepareVolume] Unable to create volume size of 105906176 bytes: device /dev/loop17p1 is not empty. Existing FS type - ext3, uuid - cb789d13-5e18-4e1e-866d-ead0d957d2b3. Set volume status to Failed
Mar 25 08:55:47.546 [INFO] [KubeClient] [UpdateCR] [pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978] Updating CR '': &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978 GenerateName: Namespace:default SelfLink: UID:24bc98fc-61bb-4238-8aca-3b9de8b8432b ResourceVersion:8430 Generation:1 CreationTimestamp:2022-03-25 08:55:38 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[app:nginx app.kubernetes.io/name:nginx] Annotations:map[] OwnerReferences:[] Finalizers:[dell.emc.csi/volume-cleanup] ClusterName: ManagedFields:[{Manager:controller Operation:Update APIVersion:csi-baremetal.dell.com/v1 Time:2022-03-25 08:55:38 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:labels":{".":{},"f:app":{},"f:app.kubernetes.io/name":{}}},"f:spec":{".":{},"f:CSIStatus":{},"f:Health":{},"f:Id":{},"f:Location":{},"f:LocationType":{},"f:Mode":{},"f:NodeId":{},"f:OperationalStatus":{},"f:Size":{},"f:StorageClass":{},"f:Type":{},"f:Usage":{}}} Subresource:} {Manager:node Operation:Update APIVersion:csi-baremetal.dell.com/v1 Time:2022-03-25 08:55:38 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{".":{},"v:\"dell.emc.csi/volume-cleanup\"":{}}}} Subresource:}]} Spec:{Id:pvc-e91c07cf-60ef-4c3d-bdff-58e30594b978 Location:73bd4b2c-fdb7-4b1d-b9e0-6869b73cca4d LocationType:DRIVE StorageClass:HDD NodeId:0e7f8009-62bd-45f4-a66e-38b09566ef66 Owners:[] Size:105906176 Mode:FS Type:xfs Health:GOOD OperationalStatus:OPERATIVE CSIStatus:FAILED Usage:IN_USE Ephemeral:false XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}}
```

### CODE REVIEW

**Comment by mishoyama** [comment]
@safronovD will there changes affect upgrade from previous CSI version where FS UUID for volumes is not set?

**Review by mishoyama — APPROVED** [review]

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 1]
```diff
@@ -312,3 +319,20 @@ func (h *WrapFSImpl) GetFSType(device string) (string, error) {
...
+// GetFSUUID detect FS UUID from the provided device using lsblk --output UUID
+// Receives file path of the device as a string
+// Returns error if something went wrong
+func (h *WrapFSImpl) GetFSUUID(device string) (string, error) {
+	var (
+		cmd    = fmt.Sprintf(GetFsUUIDCmdTmpl, device)
+		stdout string
+		err    error
+	)
+	if stdout, _, err = h.e.RunCmd(cmd,
```
Comment (rkiselenko) [inline pkg/base/linuxutils/fs/fs_helper.go:332]: I think `stderr` must be logged somewhere.

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 2]
```diff
@@ -312,3 +319,20 @@ func (h *WrapFSImpl) GetFSType(device string) (string, error) {
...
+		command.UseMetrics(true),
+		command.CmdName(fmt.Sprintf(GetFsUUIDCmdTmpl, ""))); err != nil {
```
Comment (rkiselenko) [inline pkg/base/linuxutils/fs/fs_helper.go:334]: It returns empty string template `"lsblk %s --output UUID --noheadings"` ?

**Review by rkiselenko — CHANGES_REQUESTED** [review]

**Review by mitsaa — APPROVED** [review]

File: pkg/base/linuxutils/fs/fs_helper.go (hunk 1 above)
Comment (safronovD) [reply pkg/base/linuxutils/fs/fs_helper.go:332]: `err` variable contains this info already - https://github.com/dell/csi-baremetal/blob/4b481fd70238450aceeed6efb02d333955bb8ef0/pkg/base/command/cmdexec.go#L174

File: pkg/base/linuxutils/fs/fs_helper.go (hunk 2 above)
Comment (safronovD) [reply pkg/base/linuxutils/fs/fs_helper.go:334]: Yes, for metrics selection in prometheus lib

**Comment by safronovD** [comment]
> @safronovD will there changes affect upgrade from previous CSI version where FS UUID for volumes is not set?

I expect, that it will have no impact for upgrade. We make changes only in CREATING->CREATED transition, so if volumes are already created and in usage, CSI wouldn't check them FS UUID anymore

But we can add volumes restart/deletion operations to CSI upgrade test scenario to verify it

**Review by rkiselenko — APPROVED** [review]

**Pull request merged**